blockchain/state_manager/src/lib.rs

```diff
@@ -887,4 +888,42 @@ where
         ps.miner_nominal_power_meets_consensus_minimum(self.blockstore(), addr)
             .map_err(|e| e.to_string())
     }
+
+    pub async fn validate_chain<V: ProofVerifier>(
+        &self,
+        ts: Tipset,
+    ) -> Result<(), Box<dyn StdError>> {
+        let mut ts_chain: Vec<Tipset> = vec![ts.clone()];
+        let mut ts = ts;
+        while ts.epoch() != 0 {
+            let next = chain::tipset_from_keys(self.blockstore(), ts.parents())?;
+            ts_chain.push(next.clone());
+            ts = next;
+        }
```

Comment:
```suggestion
        let mut ts_chain: Vec<Tipset> = vec![ts];
        while ts_chain.last().unwrap().epoch() != 0 {
            let last = ts_chain.last().unwrap();
            let next = chain::tipset_from_keys(self.blockstore(), last.parents())?;
            ts_chain.push(next);
        }
```
Can be cleaned up more, but this avoids a lot of clones (cloning every tipset as a duplicate just to reference is unnecessary)

Reply:
Another option is to keep the `ts` variable, start out with an empty vector, and only add the last one after the loop. Something like
```rust
let mut ts_chain = Vec::<Tipset>::new();
let mut ts = ts;
while ts.epoch() != 0 {
    let next = chain::tipset_from_keys(self.blockstore(), ts.parents())?;
    ts_chain.push(mem::replace(&mut ts, next));
}
ts_chain.push(ts);
```